### ISSUE TITLE
Add query param to `/import` endpoint to enable integration with other tools

### ISSUE THREAD

**Issue by leonardoarcari** [leonardoarcari/dcapal#34]
### Action items

- Add `?p` query param to consume JSON portfolio from `/import` endpoint
- Add formalized JSON Schema

### References

Suggested by /u/FuocoFatuo on Reddit ([link](https://www.reddit.com/r/ItaliaPersonalFinance/comments/166l6ct/comment/jyl861f/?utm_source=share&utm_medium=web2x&context=3))

**Comment by KristianLentino99** [comment]
Ciao. Ho una domanda su questo task. Mettendo un JSON in query parameter non c'è il rischio che superi la soglia massima dei query parameter? io fare un endpoint `POST` che accetta nel body della richiesta il JSON.

**Comment by leonardoarcari** [comment]
> Hello. I have a question about this task. By placing a JSON in the query parameter, is there a risk of exceeding the maximum threshold for query parameters?

Aha! Interesting I was not aware of that limit. The problem is that portfolios are all handles on FE and there's no server side rendering. So the only way to import a JSON I guess it's via query param. 

We could get away with that by using Base64 encoded query param but still the limit be hit 🤔 What do you think? 

The idea behind this task was to enable interaction with other services which could build a JSON and automatically import it into DcaPal.

**Comment by corradopetrelli** [comment]
@leonardoarcari what do you think about this package? https://pieroxy.net/blog/pages/lz-string/index.html

According to me we can use: `compressToEncodedURIComponent` and `decompressFromEncodedURIComponent` methods but probably this is not sufficient.

I have just tried a portfolio with 10 assets and there are more than 2000 characters

**Comment by KristianLentino99** [comment]
@leonardoarcari its not clear to me why you can't pass a JSON body to the import endpoint. The /import endpoint is FE side?
Then maybe we can call a simple "fake" BE API that returns to the FE the body as a response. In this way we can avoid this problem. 

Not sure I've explained me well.

**Comment by corradopetrelli** [comment]
@KristianLentino99 

As you can see [here](https://github.com/leonardoarcari/dcapal#architecture) there is not an interaction from FE to BE.
Very simply: FE fetch assets data. All calculation is FE-side with a Service Worker.

I don't know if we're talking about different things but I think this issue was opened to address the following issue: instead of having a json that you can then import into dcapal.com, Leonardo want a way to pass the json directly into the FE (throught a query string but there is the limits that you showed).

**Comment by KristianLentino99** [comment]
Ok I see, now the problem is clearer. Thanks for the clarification.

**Comment by leonardoarcari** [comment]
> I don't know if we're talking about different things but I think this issue was opened to address the following issue: instead of having a json that you can then import into dcapal.com, Leonardo want a way to pass the json directly into the FE (throught a query string but there is the limits that you showed).

Yep, that's right. Unfortunately, as you guys mentioned, this doesn't look like a viable solution. Even if compressed, having an unpredictable limit on JSON size is a terrible UX. I think this task just got bit bigger than I thought. Building on top of what @KristianLentino99 suggested we might need to implement something as follows:

- Add a `POST /api/import` endpoint to BE, validating and storing JSON temporarily in Redis (size limit, evicted in < 30 secs)
- Endpoint attaches a unique `id` to the JSON file and redirects client to some `/import?p=<id>` FE route
- FE fetches the JSON and runs the existing code to import the portfolio, load prices etc.

Ofc the new BE endpoint is put behind rate limiting in Cloudflare. Think this should scale well, enabling third-party integrations while still preventing malicious users to DOS our BE.

**Comment by KristianLentino99** [comment]
@leonardoarcari Glad to help with this change FE side. BE side I would like to help but I'm not an Rust expert unfortunately.

**Comment by corradopetrelli** [comment]
> I'm not an Rust expert unfortunately.

We will learn 😎
